basic.go

```diff
@@ -21,15 +32,486 @@ func (cb *Cinnabot) Echo(msg *message) {
+func distanceBetween(Loc1 tgbotapi.Location, Loc2 tgbotapi.Location) float64 {
+	x := math.Pow((float64(Loc1.Latitude - Loc2.Latitude)), 2)
+	y := math.Pow((float64(Loc1.Longitude - Loc2.Longitude)), 2)
+	return x + y
+}
+
+//Broadcast broadcasts a message after checking for admin status [trial]
+//Admins are to first send a message with tags before sending actual message
+func (cb *Cinnabot) Broadcast(msg *message) {
+	val := checkAdmin(cb, msg)
+	if !val {
+		cb.SendTextMessage(int(msg.Chat.ID), "🤖: Im sorry! You do not seem to be one of my overlords")
+		return
+	}
+
+	if len(msg.Args) == 0 {
+		text := "🤖: Please do /broadcast <tag>\n*Tags:*\n"
+		for i := 0; i < len(cb.allTags); i += 2 {
+			text += cb.allTags[i] + "\n"
+		}
+		cb.SendTextMessage(int(msg.Chat.ID), text)
+		return
+	}
+	//Used to initialize tags in a mark-up. Ensure that people check their tags
+	if msg.ReplyToMessage == nil {
+		//Scan for tags
+		r := regexp.MustCompile(`\/\w*`)
+		locReply := r.FindStringIndex(msg.Text)
+		tags := strings.Fields(strings.ToLower(msg.Text[locReply[1]:]))
+
+		//Filter for valid tags
+		var checkedTags []string
+		for i := 0; i < len(tags); i++ {
+			if cb.db.CheckTagExists(int(msg.Chat.ID), tags[i]) {
+				checkedTags = append(checkedTags, tags[i])
+			}
+		}
+
+		if len(checkedTags) == 0 {
+			cb.SendTextMessage(int(msg.Chat.ID), "🤖: No valid tags found")
+			return
+		}
+
+		//Send in mark-up
+		replyMsg := tgbotapi.NewMessage(msg.Chat.ID, "/broadcast "+strings.Join(checkedTags, " "))
+		replyMsg.BaseChat.ReplyToMessageID = msg.MessageID
+		replyMsg.ReplyMarkup = tgbotapi.ForceReply{ForceReply: true, Selective: true}
+		cb.SendMessage(replyMsg)
+		return
+
+	}
+
+	//Tags to send to
+	r := regexp.MustCompile(`\/\w*`)
+	locReply := r.FindStringIndex(msg.ReplyToMessage.Text)
+	tags := strings.Fields(msg.ReplyToMessage.Text[locReply[1]:])
+
+	userGroup := cb.db.UserGroup(tags)
+
+	//Forwards message to everyone in the group
+	for j := 0; j < len(userGroup); j++ {
+		forwardMess := tgbotapi.NewForward(int64(userGroup[j].UserID), msg.Chat.ID, msg.MessageID)
+		cb.SendMessage(forwardMess)
+	}
+
+	return
+}
+
+func checkAdmin(cb *Cinnabot, msg *message) bool {
+	for _, admin := range cb.keys.Admins {
+		if admin == msg.From.ID {
+			return true
+		} else if admin == int(msg.Chat.ID) {
+			return true
+		}
+	}
+	return false
+}
+
+func (cb *Cinnabot) CBS(msg *message) {
+	//Consider sending an image?
+	listText := "🤖: The Cinnamon Broadcast System (CBS) is your one stop shop for information in USP! Subscribe to the tags you want to receive notifications about!\n" +
+		"These are the following commands that you can use:\n" +
+		"/subscribe : to subscribe to a tag\n" +
+		"/unsubscribe : to unsubscribe from a tag\n\n" +
+		"*Subscribe status*\n" + "(sub status) tag: description\n"
+	for i := 0; i < len(cb.allTags); i += 2 {
+		if cb.db.CheckSubscribed(msg.From.ID, cb.allTags[i]) {
+			listText += "✅" + cb.allTags[i] + " : " + cb.allTags[i+1] + "\n"
```

Comment:
space after the emoji would be great!

Reply:
to me, a green cross is not the best for unsubscribed status. What about the following pairs:

a/ ✅❌
b/ ✅✖️
c/ ✔️✖️